File: bountydesk/validate.py
from __future__ import annotations

import re
from dataclasses import dataclass

REQUIRED_INBOX = ("opened", "platform", "title", "link", "amount", "due", "isomorphic", "decision")
REQUIRED_LEDGER = ("date", "platform", "title", "link", "status", "amount")
PROPOSAL_HEADINGS = ("problem", "do", "dont", "demo", "milestones")
MAX_OPEN_INBOX = 3
OPEN_DECISIONS = {"", "considering", "accepted", "in-progress"}
LEDGER_STATUS = {"submitted", "paid", "rejected"}

FORBIDDEN_SUBSTR = ("private_key", "privkey", "mnemonic", "seed", "wif", "secret_key")


@dataclass
class RowError:
    where: str
    message: str


def _norm(name: str) -> str:
    return "".join(ch for ch in name.lower() if ch.isalnum() or ch == "_")


def secret_keys(row: dict) -> list[str]:
    return [k for k in row if any(s in _norm(k) for s in FORBIDDEN_SUBSTR)]


def parse_md_table(text: str) -> list[dict]:
    lines = [ln.rstrip() for ln in text.splitlines() if ln.strip().startswith("|")]
    if len(lines) < 2:
        return []
    headers = [c.strip() for c in lines[0].strip("|").split("|")]
    # skip separator
    rows = []
    for ln in lines[2:]:
        cells = [c.strip() for c in ln.strip("|").split("|")]
        if len(cells) != len(headers):
            continue
        row = dict(zip(headers, cells))
        if all(v == "" for v in row.values()):
            continue
        rows.append(row)
    return rows


def validate_inbox(rows: list[dict]) -> list[RowError]:
    errs: list[RowError] = []
    for i, row in enumerate(rows, start=1):
        for key in secret_keys(row):
            errs.append(RowError(f"inbox[{i}]", f"forbidden field {key}"))
        for col in REQUIRED_INBOX:
            if col not in row:
                errs.append(RowError(f"inbox[{i}]", f"missing {col}"))
        if row.get("link") and not re.match(r"https?://", row["link"]):
            errs.append(RowError(f"inbox[{i}]", "link must be http(s)"))
    open_rows = [r for r in rows if r.get("decision", "").lower() in OPEN_DECISIONS]
    if len(open_rows) > MAX_OPEN_INBOX:
        errs.append(RowError("inbox", f"max {MAX_OPEN_INBOX} open rows, have {len(open_rows)}"))
    return errs


def validate_ledger(rows: list[dict]) -> list[RowError]:
    errs: list[RowError] = []
    for i, row in enumerate(rows, start=1):
        for key in secret_keys(row):
            errs.append(RowError(f"ledger[{i}]", f"forbidden field {key}"))
        for col in REQUIRED_LEDGER:
            if col not in row:
                errs.append(RowError(f"ledger[{i}]", f"missing {col}"))
        st = row.get("status", "")
        if st and st not in LEDGER_STATUS:
            errs.append(RowError(f"ledger[{i}]", f"bad status {st}"))
    return errs


def _heading_tokens(text: str) -> set[str]:
    """First word of each markdown heading line. '## dont' is only 'dont', not 'do'."""
    found: set[str] = set()
    for raw in text.splitlines():
        line = raw.strip()
        if not line.startswith("#"):
            continue
        title = line.lstrip("#").strip().lower()
        if not title:
            continue
        found.add(title.split()[0])
    return found


def validate_proposal(text: str) -> list[RowError]:
    present = _heading_tokens(text)
    errs = []
    for h in PROPOSAL_HEADINGS:
        if h not in present:
            errs.append(RowError("proposal", f"missing heading {h}"))
    if "private_key" in text.lower():
        errs.append(RowError("proposal", "looks like it contains a private_key"))
    return errs

File: src/validate.rs
use regex::Regex;
use std::collections::{HashMap, HashSet};

pub const MAX_OPEN_INBOX: usize = 3;
const REQUIRED_INBOX: &[&str] = &[
    "opened",
    "platform",
    "title",
    "link",
    "amount",
    "due",
    "isomorphic",
    "decision",
];
const REQUIRED_LEDGER: &[&str] = &["date", "platform", "title", "link", "status", "amount"];
const PROPOSAL_HEADINGS: &[&str] = &["problem", "do", "dont", "demo", "milestones"];
const FORBIDDEN: &[&str] = &["private_key", "privkey", "mnemonic", "seed", "wif", "secret_key"];

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct RowError {
    pub where_: String,
    pub message: String,
}

impl std::fmt::Display for RowError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}: {}", self.where_, self.message)
    }
}

fn norm(name: &str) -> String {
    name.chars()
        .filter(|c| c.is_ascii_alphanumeric() || *c == '_')
        .flat_map(|c| c.to_lowercase())
        .collect()
}

fn secret_keys(row: &HashMap<String, String>) -> Vec<String> {
    row.keys()
        .filter(|k| FORBIDDEN.iter().any(|n| norm(k).contains(n)))
        .cloned()
        .collect()
}

pub fn parse_md_table(text: &str) -> Vec<HashMap<String, String>> {
    let lines: Vec<&str> = text
        .lines()
        .filter(|ln| ln.trim_start().starts_with('|'))
        .collect();
    if lines.len() < 2 {
        return Vec::new();
    }
    let headers: Vec<String> = lines[0]
        .trim()
        .trim_matches('|')
        .split('|')
        .map(|c| c.trim().to_string())
        .collect();
    let mut rows = Vec::new();
    for ln in lines.iter().skip(2) {
        let cells: Vec<String> = ln
            .trim()
            .trim_matches('|')
            .split('|')
            .map(|c| c.trim().to_string())
            .collect();
        if cells.len() != headers.len() {
            continue;
        }
        let row: HashMap<_, _> = headers.iter().cloned().zip(cells).collect();
        if row.values().all(|v| v.is_empty()) {
            continue;
        }
        rows.push(row);
    }
    rows
}

pub fn validate_inbox(rows: &[HashMap<String, String>]) -> Vec<RowError> {
    let mut errs = Vec::new();
    let http = Regex::new(r"^https?://").unwrap();
    for (i, row) in rows.iter().enumerate() {
        let n = i + 1;
        for key in secret_keys(row) {
            errs.push(RowError {
                where_: format!("inbox[{n}]"),
                message: format!("forbidden field {key}"),
            });
        }
        for col in REQUIRED_INBOX {
            if !row.contains_key(*col) {
                errs.push(RowError {
                    where_: format!("inbox[{n}]"),
                    message: format!("missing {col}"),
                });
            }
        }
        if let Some(link) = row.get("link") {
            if !link.is_empty() && !http.is_match(link) {
                errs.push(RowError {
                    where_: format!("inbox[{n}]"),
                    message: "link must be http(s)".into(),
                });
            }
        }
    }
    let open: Vec<_> = rows
        .iter()
        .filter(|r| {
            matches!(
                r.get("decision").map(|s| s.to_lowercase()).as_deref(),
                Some("") | Some("considering") | Some("accepted") | Some("in-progress") | None
            )
        })
        .collect();
    if open.len() > MAX_OPEN_INBOX {
        errs.push(RowError {
            where_: "inbox".into(),
            message: format!("max {MAX_OPEN_INBOX} open rows, have {}", open.len()),
        });
    }
    errs
}

pub fn validate_ledger(rows: &[HashMap<String, String>]) -> Vec<RowError> {
    let mut errs = Vec::new();
    for (i, row) in rows.iter().enumerate() {
        let n = i + 1;
        for key in secret_keys(row) {
            errs.push(RowError {
                where_: format!("ledger[{n}]"),
                message: format!("forbidden field {key}"),
            });
        }
        for col in REQUIRED_LEDGER {
            if !row.contains_key(*col) {
                errs.push(RowError {
                    where_: format!("ledger[{n}]"),
                    message: format!("missing {col}"),
                });
            }
        }
        if let Some(st) = row.get("status") {
            if !st.is_empty() && !matches!(st.as_str(), "submitted" | "paid" | "rejected") {
                errs.push(RowError {
                    where_: format!("ledger[{n}]"),
                    message: format!("bad status {st}"),
                });
            }
        }
    }
    errs
}

pub fn heading_tokens(text: &str) -> HashSet<String> {
    let mut found = HashSet::new();
    for raw in text.lines() {
        let line = raw.trim();
        if !line.starts_with('#') {
            continue;
        }
        let title = line.trim_start_matches('#').trim().to_lowercase();
        if let Some(first) = title.split_whitespace().next() {
            found.insert(first.to_string());
        }
    }
    found
}

pub fn validate_proposal(text: &str) -> Vec<RowError> {
    let present = heading_tokens(text);
    let mut errs = Vec::new();
    for h in PROPOSAL_HEADINGS {
        if !present.contains(*h) {
            errs.push(RowError {
                where_: "proposal".into(),
                message: format!("missing heading {h}"),
            });
        }
    }
    if text.to_lowercase().contains("private_key") {
        errs.push(RowError {
            where_: "proposal".into(),
            message: "looks like it contains a private_key".into(),
        });
    }
    errs
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn max_three_open() {
        let mut rows = Vec::new();
        for i in 0..4 {
            let mut r = HashMap::new();
            r.insert("opened".into(), "2026-08-18".into());
            r.insert("platform".into(), "gitcoin".into());
            r.insert("title".into(), format!("t{i}"));
            r.insert("link".into(), "https://example.com".into());
            r.insert("amount".into(), "1".into());
            r.insert("due".into(), "x".into());
            r.insert("isomorphic".into(), "yes".into());
            r.insert("decision".into(), "considering".into());
            rows.push(r);
        }
        let errs = validate_inbox(&rows);
        assert!(errs.iter().any(|e| e.message.contains("max 3")));
    }

    #[test]
    fn missing_do_not_satisfied_by_dont() {
        let text = "## problem\nbody\n## dont\nno keys\n## demo\nlink\n## milestones\none\n";
        let errs = validate_proposal(text);
        assert!(errs.iter().any(|e| e.message == "missing heading do"));
        assert!(!errs.iter().any(|e| e.message.contains("missing heading dont")));
    }

    #[test]
    fn example_proposal_ok() {
        let text = std::fs::read_to_string(
            std::path::Path::new(env!("CARGO_MANIFEST_DIR")).join("docs/proposal.example.md"),
        )
        .unwrap();
        assert!(validate_proposal(&text).is_empty());
    }
}
